packages/core/http/core-http-router-server-internal/src/versioned_router/oas_poc/generate_oas.ts

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/**
+ * Heavily adapted version of https://github.com/jlalmes/trpc-openapi
+ */
+
+/* eslint-disable import/no-extraneous-dependencies */
+
+import type { OpenAPIV3 } from 'openapi-types';
+import zodToJsonSchema from 'zod-to-json-schema';
+import {
+  instanceofZodType,
+  instanceofZodTypeObject,
+  instanceofZodTypeLikeVoid,
+  instanceofZodTypeLikeString,
+  instanceofZodTypeOptional,
+  unwrapZodType,
+  zodSupportsCoerce,
+  instanceofZodTypeCoercible,
+  getPathParameters,
+} from './util';
+
+import type { CoreVersionedRouter } from '..';
+import type { VersionedRouterRoute } from '../types';
+
+export const openApiVersion = '3.0.0';
+
+export interface GenerateOpenApiDocumentOptions {
+  title: string;
+  description?: string;
+  version: string;
+  baseUrl: string;
+  docsUrl?: string;
+  tags?: string[];
+}
+
+export function generateOpenApiDocument(
+  appRouter: CoreVersionedRouter,
+  opts: GenerateOpenApiDocumentOptions
+): OpenAPIV3.Document {
+  return {
+    openapi: openApiVersion,
+    info: {
+      title: opts.title,
+      description: opts.description,
+      version: opts.version,
+    },
+    servers: [
+      {
+        url: opts.baseUrl,
+      },
+    ],
+    paths: getOpenApiPathsObject(appRouter),
+    // components: {
+    //   securitySchemes: {
+    //     Authorization: {
+    //       type: 'http',
+    //       scheme: 'bearer',
+    //     },
+    //   },
+    //   responses: {
+    //     error: errorResponseObject,
+    //   },
+    // },
+    tags: opts.tags?.map((tag) => ({ name: tag })),
+    externalDocs: opts.docsUrl ? { url: opts.docsUrl } : undefined,
+  };
```

Comment:
One thing I don't see yet in this example is the [components object](https://spec.openapis.org/oas/v3.1.0#components-object), which is really handy for re-using common schemas, parameters, etc.

When I ran `npx @redocly/cli lint <file>.json`, it also noted that "every operation should have security defined on it or on the root level". There's more info about the security object in https://spec.openapis.org/oas/v3.1.0#securityRequirementObject. You can see an example of how we've coded it in [ML](https://github.com/elastic/kibana/blob/8c8bb8f798e1d1fdc3c72df203460a592d1221a2/x-pack/plugins/ml/common/openapi/ml_apis_v3.yaml#L186) or  [Cases](https://github.com/elastic/kibana/blob/8c8bb8f798e1d1fdc3c72df203460a592d1221a2/x-pack/plugins/cases/docs/openapi/bundled.json#L2230)

Reply:
Good point! This is where "code-first" OAS departs from "spec-first" OAS authoring. Endpoint authors write a their route components in TypeScript which gets converted into the corresponding JSONSchema. The current conversion process does not abstract out specific components (even though there may be identical schemas), at a code level they are already factored into components.

My understanding of components is they are a convenience mechanism for spec authors, but given we are deriving spec from objects in code I thought we might not need to generate components. Does that understanding of components track? I may certainly be missing something!

---

> every operation should have security defined on it or on the root level

That's a great point too. We could just include a "dumb" basic auth value on all objects automatically and then refine the ability to specify different security values later. I'll look into this!

Reply:
Yes, I think components are a convenience, but also for users who can see in the end product that object XYZ gets used in multiple endpoints (e.g. the same info going into a put operation and coming out of a get). I think it's okay if they're not possible, though.